### ISSUE TITLE
Screengrab crashes on Android 7+

### ISSUE THREAD

**Issue by SemenovAlexander** [fastlane/fastlane#7592]
Screengrab uses deprecated file access mode here
https://github.com/fastlane/fastlane/blob/master/screengrab/screengrab-lib/src/main/java/tools.fastlane.screengrab/FileWritingScreenshotCallback.java#L70

obviously, run on Android 7+ device causes Screengrab to crash with 
`java.lang.SecurityException: MODE_WORLD_READABLE no longer supported`

See details:
https://developer.android.com/reference/android/content/Context.html#MODE_WORLD_READABLE

We should think for some other method to store screenshots files in Android 7, screenshots from latest os version is a critical for such tool.

**Comment by hjanuschka** [comment]
@SemenovAlexander can you checkout the PR here -> https://github.com/fastlane/fastlane/issues/7616

according to the docs changing WORLD_READABLE to PRIVATE fixes it - can you give it a try?

**Comment by SemenovAlexander** [comment]
@hjanuschka yeap, just tried, and it works like a charm for me!

However, I have no option to check this on real device, tried only on emulator.

**Comment by mfurtak** [comment]
Hi @SemenovAlexander - thanks for the report!

I'll look into this, and @hjanuschka's proposed fix 👍

**Comment by mfurtak** [comment]
@SemenovAlexander Can you please share with me your `androidTestCompile` statement that sets up `screengrab` for your project?

Also, what is the output of `screengrab -v`?

Thanks!

**Comment by mfurtak** [comment]
@SemenovAlexander I have tried to reproduce this on both an API 25 (Android 7.1.1 Google APIs) emulator, and a Nexus 6p running Nougat. In both cases things are working without issue.

Can you please provide more context around this problem, such as the command line output from running _screengrab_ and/or any stack traces you see from the Android side?

Thanks!

**Comment by SemenovAlexander** [comment]
@mfurtak sorry for late answer, I've missed the reminder somehow. Sure:
`androidTestCompile 'tools.fastlane:screengrab:0.5.6'`
And here is output, when I call `Screengrab.screenshot("test")`:

```
java.lang.RuntimeException: Unable to capture screenshot.
at tools.fastlane.screengrab.FileWritingScreenshotCallback.screenshotCaptured(FileWritingScreenshotCallback.java:57)
at tools.fastlane.screengrab.UiAutomatorScreenshotStrategy.takeScreenshot(UiAutomatorScreenshotStrategy.java:39)
at tools.fastlane.screengrab.Screengrab.screenshot(Screengrab.java:94)
at tools.fastlane.screengrab.Screengrab.screenshot(Screengrab.java:76)
at tools.fastlane.screengrab.Screengrab.screenshot(Screengrab.java:60)
at com.vpnshieldapp.androidclient.test.utils.TestUtils.screenshot(TestUtils.java:67)
at com.vpnshieldapp.androidclient.test.AccountTest.userMail(AccountTest.java:99)
at com.vpnshieldapp.androidclient.test.AccountTest.openAccount(AccountTest.java:93)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
at android.support.test.internal.statement.UiThreadStatement.evaluate(UiThreadStatement.java:55)
at android.support.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:270)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at tools.fastlane.screengrab.locale.LocaleTestRule$1.evaluate(LocaleTestRule.java:32)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at android.support.test.internal.runner.TestExecutor.execute(TestExecutor.java:59)
at android.support.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:262)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1932)
Caused by: java.lang.SecurityException: MODE_WORLD_READABLE no longer supported
at android.app.ContextImpl.checkMode(ContextImpl.java:2134)
at android.app.ContextImpl.getDir(ContextImpl.java:1955)
at android.content.ContextWrapper.getDir(ContextWrapper.java:278)
at tools.fastlane.screengrab.FileWritingScreenshotCallback.getFilesDirectory(FileWritingScreenshotCallback.java:70)
at tools.fastlane.screengrab.FileWritingScreenshotCallback.screenshotCaptured(FileWritingScreenshotCallback.java:39)
... 40 more
```

**Comment by mfurtak** [comment]
Aha, thanks! That is helpful. Do you mind sending me the details for the emulator you are using?

![image](https://cloud.githubusercontent.com/assets/343134/21868144/6cf71f60-d81f-11e6-8bae-2db560361cf1.png)

![image](https://cloud.githubusercontent.com/assets/343134/21868136/66925f04-d81f-11e6-812a-0bb770fed235.png)

**Comment by SemenovAlexander** [comment]
Sure, see attachment.

It seems, directory is not null here in your case and is null in mine.
https://github.com/fastlane/fastlane/blob/master/screengrab/screengrab-lib/src/main/java/tools.fastlane.screengrab/FileWritingScreenshotCallback.java#L73
 
<img width="1229" alt="screen shot 2017-01-12 at 00 03 17" src="https://cloud.githubusercontent.com/assets/6439273/21868271/c00d2a92-d85a-11e6-891a-568fda557058.png">

**Comment by mfurtak** [comment]
Hm, yes, I agree!

I was hoping I could explain it by seeing whether your emulator had SD Card storage available. It seems like it does, so there must be some other reason that the directory on external storage is failing to be initialized.

Is the SD card full (no free space), by any chance?

**Comment by SemenovAlexander** [comment]
Nope, sd card is fine, I've attached debugger, and issue is 
`dir.canWrite()` returns false for first constructed path
`/storage/emulated/0/com.vpnshieldapp/screengrab/en-US/images/screenshots`

Ah, it's my fault, the issue is, I was running tests directly from Android Studio, when faced that issue, changed FireWritingCallback and forgot about that. 

Running via Screengrab all works fine, because app is granted `WRITE_EXTERNAL_STORAGE` and first check here will pass.
https://github.com/fastlane/fastlane/blob/master/screengrab/screengrab-lib/src/main/java/tools.fastlane.screengrab/FileWritingScreenshotCallback.java#L69 

However, I think #7616 should still be accepted, as that code will crash on Nougat if execution reaches it for any reason.
https://github.com/fastlane/fastlane/blob/master/screengrab/screengrab-lib/src/main/java/tools.fastlane.screengrab/FileWritingScreenshotCallback.java#L74

**Comment by mfurtak** [comment]
@SemenovAlexander Aha! OK, great. I'm glad you are back in business. :rocket:

I am going to prepare a change to prevent this failure mode for Android N. I don't quite agree with the change in #7616, because using this file mode in the fall-back scenario is important for letting things work on older devices. So, going forward this check will be like:

```java
if (directory == null && Build.VERSION.SDK_INT < 24) {
    File internalDir = new File(context.getDir(SCREENGRAB_DIR_NAME, Context.MODE_WORLD_READABLE), localeToDirName(locale));
    directory = initializeDirectory(internalDir);
}
```

which should prevent this particular failure mode in the future.

**Comment by mfurtak** [comment]
This is released in the [v1.0.1](https://bintray.com/fastlane/fastlane/screengrab/1.0.1) of the AAR 🚀
